src/passes/AbstractTypeRefining.cpp

```diff
@@ -297,11 +302,69 @@ struct AbstractTypeRefining : public Pass {
 
     AbstractTypeRefiningTypeMapper(*module, mapping).map();
 
+    if (optimizedDescribedType) {
+      // At this point we may have casts like this:
+      //
+      // (ref.cast_desc nullref
+      //   (some struct...)
+      //   (some desc...)
+      // )
+      //
+      // ReFinalize would fix up the cast target to be the struct type described
+      // by the descriptor, but that struct type was supposed to have been
+      // optimized out. Optimize out the cast (which we know must either be a
+      // null check or unconditional trap) before ReFinalize can get to it.
```

Comment:
Can we perhaps modify ReFinalize instead? If it sees `castType` is nullref, it can leave it that way, to avoid undoing work?

Reply:
It's never valid to have a descriptor cast to null, so we would have to do this same fixup one way or another. It makes more sense to have the fixup in AbstractTypeRefining, which uniquely introduces the problem, rather than a more general-purpose place like ReFinalize.